# Boat Sales Analysis

### Part 1 - Data Cleaning & Data Wrangling

#### Content
1. Import Libraries
2. Load data set 
3. Explore data content
4. Data cleaning
5. Data wrangling
6. Save data set into a file

#### 1. Import libraries

In [1]:
import pandas as pd
import numpy as np
import os
import seaborn as sns
import matplotlib.pyplot as plt 

#### 2. Load Data Set

In [2]:
# Set path to access data files
path = r'C:\Users\tsoew\OneDrive\Desktop\Boat Sales Analysis'

In [3]:
# Import file into dataframe
df = pd.read_csv (os.path.join(path, 'Data', 'Original Data', 'boat_data_orig.csv'), index_col=False)

In [4]:
# Check data head
df.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

#### Note:
- Location column has some garbled characters which could be a divider between country name and city name.

In [5]:
df.shape

(9888, 10)

In [5]:
df['Year Built'].unique

<bound method Series.unique of 0       2017
1       2020
2          0
3       2020
4       2019
        ... 
9883    1987
9884       0
9885    2020
9886    2018
9887    2019
Name: Year Built, Length: 9888, dtype: int64>

#### 3. Explore Data content

In [6]:
df.describe()

Year Built       Length        Width  Number of views last 7 days
count  9888.000000  9879.000000  9832.000000                  9888.000000
mean   1893.192860    11.570017     3.520124                   149.160801
std     460.201582     6.002820     1.220534                   151.819752
min       0.000000     1.040000     0.010000                    13.000000
25%    1996.000000     7.470000     2.540000                    70.000000
50%    2007.000000    10.280000     3.330000                   108.000000
75%    2017.000000    13.930000     4.250000                   172.000000
max    2021.000000   100.000000    25.160000                  3263.000000

#### Note for some anomalies:
- Year Built minimum has 0 values and average of 1893, which is quite outdated.
- Length minimum is 1.04 which is too small for a boat.
- Width minimum is 0.01 which is not possible.
- Num of Views has maximum of 3263 which can be an outlier compared to other values.

In [7]:
df.shape

(9888, 10)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Price                        9888 non-null   object 
 1   Boat Type                    9888 non-null   object 
 2   Manufacturer                 8550 non-null   object 
 3   Type                         9882 non-null   object 
 4   Year Built                   9888 non-null   int64  
 5   Length                       9879 non-null   float64
 6   Width                        9832 non-null   float64
 7   Material                     8139 non-null   object 
 8   Location                     9852 non-null   object 
 9   Number of views last 7 days  9888 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 772.6+ KB


### 4. Data Cleaning

#### Data cleaning: "Year Built" column

In [9]:
df['Year Built'].value_counts()

Year Built
2020    1279
2019     663
0        551
2008     457
2007     393
        ... 
1914       1
1895       1
1885       1
1931       1
1900       1
Name: count, Length: 122, dtype: int64

In [10]:
# Remove 0 values in 'Year Built' column
df = df[df['Year Built'] != 0]

In [11]:
df.shape

(9337, 10)

#### Data cleaning: "Width" column

In [12]:
# remove any boat with width under 1 meter
df = df[df['Width']>=1]

In [13]:
df.shape

(9276, 10)

In [14]:
# fill null values with average
df['Width'] = df['Width'].fillna(df['Width'].mean())

#### Data cleaning: "Length" column

In [15]:
# remove any boat with length under 1.5 meter
df = df[df['Length']>=1.5]

In [16]:
df.shape

(9274, 10)

In [17]:
# fill null values with average
df['Length'] = df['Length'].fillna(df['Length'].mean())

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9274 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Price                        9274 non-null   object 
 1   Boat Type                    9274 non-null   object 
 2   Manufacturer                 8007 non-null   object 
 3   Type                         9269 non-null   object 
 4   Year Built                   9274 non-null   int64  
 5   Length                       9274 non-null   float64
 6   Width                        9274 non-null   float64
 7   Material                     7647 non-null   object 
 8   Location                     9238 non-null   object 
 9   Number of views last 7 days  9274 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 797.0+ KB


#### Data cleaning: 'Number of views last 7 days' column

In [19]:
#rename column for simplicity
df.rename(columns={"Number of views last 7 days": "Views"}, inplace=True)

In [20]:
df[df['Views']>2000]

Price                        Boat Type          Manufacturer  \
308    CHF 19900                       Sport Boat   Sea Ray power boats   
3700  EUR 949000                          Hardtop  Pershing power boats   
6211  CHF 125900  Bowrider,Motor Yacht,Sport Boat     Windy power boats   
8723   CHF 35000                          Hardtop  Princess power boats   
9580   CHF 14900                      Motor Yacht  Bayliner power boats   

                    Type  Year Built  Length  Width Material  \
308   Used boat,Unleaded        1993    6.14   2.34  Plastic   
3700    Used boat,Diesel        2009   20.30   5.20      GRP   
6211    Used boat,Diesel        2002   12.35   3.48      GRP   
8723    Used boat,Diesel        1979   11.12   3.88      GRP   
9580  Used boat,Unleaded        1992    7.70   2.46  Plastic   

                                           Location  Views  
308            Switzerland » Murtensee » Avenches   2026  
3700                  Neustadt in Holstein (Ostsee)   2154  
6211  Switzerland » Lago Maggiore » 6600 Locarno    2261  
8723     Switzerland » Neuenburgersee » Hauterive   2432  
9580                Switzerland » Le Landeron (NE)   3263

#### Note:
It looks like the number 3263 is not an outlier, but it seems possible for a very popular boat. So we will leave this column alone.

#### Remove columns that are not used for this analysis

In [21]:
# Remove 'Manufacturer' column
df.drop('Manufacturer', axis=1, inplace=True)

In [22]:
# Remove 'Material' column
df.drop('Material', axis=1, inplace=True)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9274 entries, 0 to 9887
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Price       9274 non-null   object 
 1   Boat Type   9274 non-null   object 
 2   Type        9269 non-null   object 
 3   Year Built  9274 non-null   int64  
 4   Length      9274 non-null   float64
 5   Width       9274 non-null   float64
 6   Location    9238 non-null   object 
 7   Views       9274 non-null   int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 652.1+ KB


In [24]:
# Drop all null values
df.dropna(inplace=True)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9233 entries, 0 to 9887
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Price       9233 non-null   object 
 1   Boat Type   9233 non-null   object 
 2   Type        9233 non-null   object 
 3   Year Built  9233 non-null   int64  
 4   Length      9233 non-null   float64
 5   Width       9233 non-null   float64
 6   Location    9233 non-null   object 
 7   Views       9233 non-null   int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 649.2+ KB


### 5. Data Wrangling

#### Reformat Price column

In [26]:
#Split Price to 2 columns: 'currency' & 'Price'
df[['currency','Price']]= df['Price'].str.split(' ', expand=True)

In [27]:
# Convert all prices into US Dollar currency

def USDprice(Price,currency):
    
    if currency=='EUR':              
        #Euro
        return int(Price) * 1.075
    elif currency=='CHF':            
        #Swiss Franc
        return int(Price) * 1.099
    elif currency=='DKK':            
        #Danish Krone
        return int(Price) * 0.143
    elif currency=='£':             
        #British Pounds
        return int(Price) * 1.25
    
df['Price_USD'] = df.apply(lambda x: USDprice(x.Price, x.currency), axis=1)

In [28]:
# delete Price and currency columns as we are not using them anymore
df.drop('currency', axis=1, inplace=True)
df.drop('Price', axis=1, inplace=True)

In [29]:
df.head()

Boat Type                 Type  Year Built  Length  Width  \
0          Motor Yacht  new boat from stock        2017    4.00   1.90   
1  Center console boat  new boat from stock        2020    4.00   1.50   
3           Sport Boat  new boat from stock        2020    3.00   1.00   
4         Fishing Boat  new boat from stock        2019    3.55   1.46   
6            Catamaran   Used boat,Unleaded        1999    6.20   2.38   

                                    Location  Views  Price_USD  
0     Switzerland » Lake Geneva » Vésenaz    226   3667.363  
1                   Germany » Bönningstedt     75   3751.750  
3                       Denmark » Svendborg     64   3703.700  
4              Germany » Bayern » München     58   3653.925  
6  Switzerland » Neuenburgersee » Yvonand     474   3956.400

#### Reformat Location column

In [30]:
#Split Location to 3 columns: 'Country','City','Other'
df[['Country','City', 'Other']]= df['Location'].str.split('»', expand=True)

In [31]:
# remove some garbage characters
df['Country'] = df['Country'].str.replace(" ", "", regex=True)
df['City']=df['City'].str.replace('¶','',regex=True)
df['City']=df['City'].str.replace('Ã','',regex=True)
df['City']=df['City'].str.replace(' ','',regex=True)

In [32]:
# Drop unsed columns
df.drop('Location', axis=1, inplace=True)
df.drop('Other', axis=1, inplace=True)
df.drop('City', axis=1, inplace=True)

In [33]:
df.shape

(9233, 8)

In [34]:
df.head()

Boat Type                 Type  Year Built  Length  Width  Views  \
0          Motor Yacht  new boat from stock        2017    4.00   1.90    226   
1  Center console boat  new boat from stock        2020    4.00   1.50     75   
3           Sport Boat  new boat from stock        2020    3.00   1.00     64   
4         Fishing Boat  new boat from stock        2019    3.55   1.46     58   
6            Catamaran   Used boat,Unleaded        1999    6.20   2.38    474   

   Price_USD      Country  
0   3667.363  Switzerland  
1   3751.750      Germany  
3   3703.700      Denmark  
4   3653.925      Germany  
6   3956.400  Switzerland

### 6. Save clean data into a new file

In [35]:
# Export the new format of department dataframe
df.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'boat_data_clean.csv'))

### --------------- End of Boat Analysis Part 1 -------------------